src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -227,24 +228,99 @@ public int capacity() {
      */
     public void ensureCapacity(int minimumCapacity) {
         if (minimumCapacity > 0) {
-            ensureCapacityInternal(minimumCapacity);
+            value = ensureCapacitySameCoder(value, coder, minimumCapacity);
         }
     }
 
     /**
+     * {@return true if the byte array should be replaced due to increased capacity or coder change}
+     * <ul>
+     *     <li>The new coder is the different thn the old coder
+     *     <li>The new length is greater than to the current length
+     *     <li>The new length is non-negative, as it might have overflowed due to an increment
+     * </ul>
+     *
+     * @param value       a byte array
+     * @param coder       old coder
+     * @param newCapacity new capacity in characters
+     * @param newCoder    new coder
+     */
+    private static boolean needsNewBuffer(byte[] value, byte coder, int newCapacity, byte newCoder) {
+        long newLength = (long) newCapacity << newCoder;
+        return coder != newCoder || newLength > value.length || 0 > newLength;
+    }
+
+
+    /**
+     * {@return a buffer suitable for storing the bytes up to minimum capacity using the new coder}
+     * If the coder matches and there is enough room, the same buffer is returned.
+     * Otherwise, a new buffer is allocated and the string is copied or inflated to match the new coder.
      * For positive values of {@code minimumCapacity}, this method
-     * behaves like {@code ensureCapacity}, however it is never
-     * synchronized.
-     * If {@code minimumCapacity} is non positive due to numeric
+     * behaves like {@code ensureCapacity}, however it is never synchronized.
+     * If {@code minimumCapacity} is non-positive due to numeric
      * overflow, this method throws {@code OutOfMemoryError}.
+     * @param value the current buffer
+     * @param coder of the current buffer
+     * @param count the count of chars in the current buffer
+     * @param minimumCapacity the new minimum capacity
+     * @param newCoder the desired new coder
+     */
+    private static byte[] ensureCapacityNewCoder(byte[] value, byte coder, int count,
+                                                 int minimumCapacity, byte newCoder) {
+        assert coder == newCoder || newCoder == UTF16 : "bad new coder UTF16 -> LATIN1";
+        // overflow-conscious code
+        int growth = minimumCapacity - (value.length >> coder);
+        if (coder == newCoder) {
+            if (growth > 0) {
+                // copy all bytes to new larger buffer
+                value = Arrays.copyOf(value,
+                        newCapacity(value, newCoder, minimumCapacity) << newCoder);
```

Comment:
Would this benefit with an `assert` to verify that the new buffer length being proposed is not lesser than the current buffer's length (thus truncating content). Something like:

```
// copy all bytes to new larger buffer
int newLen = newCapacity(value, newCoder, minimumCapacity) << newCoder;
assert newLen >= value.length : "bad new length " + newLen + " for buffer's length " + value.length; 
value = Arrays.copyOf(value, newLen);
```